olap/src/aggregator/ceramic_patch.rs

```diff
@@ -0,0 +1,147 @@
+use std::sync::Arc;
+
+use arrow::{
+    array::{Array as _, ArrayBuilder as _, ArrayRef, StringBuilder},
+    datatypes::DataType,
+};
+use datafusion::{
+    common::{
+        cast::{as_binary_array, as_string_array},
+        exec_datafusion_err, Result,
+    },
+    logical_expr::{
+        PartitionEvaluator, Signature, TypeSignature, Volatility, WindowUDF, WindowUDFImpl,
+    },
+};
+
+/// Applies a Ceramic data event to a document state returning the new document state.
+#[derive(Debug)]
+pub struct CeramicPatch {
+    signature: Signature,
+}
+
+impl CeramicPatch {
+    pub fn new_udwf() -> WindowUDF {
+        WindowUDF::new_from_impl(Self::new())
+    }
+    fn new() -> Self {
+        Self {
+            signature: Signature::new(
+                TypeSignature::Exact(vec![
+                    DataType::Binary,
+                    DataType::Binary,
+                    DataType::Utf8,
+                    DataType::Utf8,
+                ]),
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl WindowUDFImpl for CeramicPatch {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "ceramic_patch"
+    }
+
+    fn signature(&self) -> &datafusion::logical_expr::Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Utf8)
+    }
+
+    fn partition_evaluator(&self) -> Result<Box<dyn PartitionEvaluator>> {
+        Ok(Box::new(CeramicPatchEvaluator))
+    }
+}
+
+#[derive(Debug)]
+struct CeramicPatchEvaluator;
+
+impl CeramicPatchEvaluator {
+    fn apply_patch(patch: &str, previous_state: &str) -> Result<String> {
+        let patch: Vec<json_patch::PatchOperation> = serde_json::from_str(patch)
+            .map_err(|err| exec_datafusion_err!("Error parsing patch: {err}"))?;
+        let mut new_state: serde_json::Value = serde_json::from_str(previous_state)
+            .map_err(|err| exec_datafusion_err!("Error parsing previous state: {err}"))?;
+        json_patch::patch(&mut new_state, &patch)
+            .map_err(|err| exec_datafusion_err!("Error applying JSON patch: {err}"))?;
+        serde_json::to_string(&new_state)
+            .map_err(|err| exec_datafusion_err!("Error JSON encoding: {err}"))
+    }
+}
+
+impl PartitionEvaluator for CeramicPatchEvaluator {
+    fn evaluate_all(&mut self, values: &[ArrayRef], num_rows: usize) -> Result<ArrayRef> {
```

Comment:
The main thing I struggled with was understanding the where things are coming from and what we're actually working with. The datafusion docs are pretty good for how this trait is used but but including a doc comment about what the expected input looks like would be helpful. Then again, the assignments into the fields makes it fairly clear, so maybe it's not useful 🤷

(these assignments)
```rust
        let event_cids = as_binary_array(&values[0])?;
        let previous_cids = as_binary_array(&values[1])?;
        let previous_states = as_string_array(&values[2])?;
        let patches = as_string_array(&values[3])?;